onmt/trainer.py

```diff
@@ -133,79 +152,60 @@ def train(self,
-                    else:
-                        normalization += batch.batch_size
-                    accum += 1
-                    if accum == self.grad_accum_count:
-                        reduce_counter += 1
-                        if self.gpu_verbose_level > 0:
-                            logger.info("GpuRank %d: reduce_counter: %d \
-                                        n_minibatch %d"
-                                        % (self.gpu_rank, reduce_counter,
-                                           len(true_batches)))
-                        if self.n_gpu > 1:
-                            normalization = sum(onmt.utils.distributed
-                                                .all_gather_list
-                                                (normalization))
-
-                        self._gradient_accumulation(
-                            true_batches, normalization, total_stats,
-                            report_stats)
-
-                        report_stats = self._maybe_report_training(
-                            step, train_steps,
-                            self.optim.learning_rate(),
-                            report_stats)
-
-                        true_batches = []
-                        accum = 0
-                        normalization = 0
-                        if valid_iter is not None and step % valid_steps == 0:
-                            if self.gpu_verbose_level > 0:
-                                logger.info('GpuRank %d: validate step %d'
-                                            % (self.gpu_rank, step))
-                            valid_stats = self.validate(valid_iter)
-                            if self.gpu_verbose_level > 0:
-                                logger.info('GpuRank %d: gather valid stat \
-                                            step %d' % (self.gpu_rank, step))
-                            valid_stats = self._maybe_gather_stats(valid_stats)
-                            if self.gpu_verbose_level > 0:
-                                logger.info('GpuRank %d: report stat step %d'
-                                            % (self.gpu_rank, step))
-                            self._report_step(self.optim.learning_rate(),
-                                              step, valid_stats=valid_stats)
-
-                        if self.gpu_rank == 0:
-                            self._maybe_save(step)
-                        step += 1
-                        if step > train_steps:
-                            break
+        if self.n_gpu > 1:
+            train_iter = itertools.islice(
+                train_iter, self.gpu_rank, None, self.n_gpu)
+
+        reduce_counter = 0
+        for i, (batches, normalization) in enumerate(
+                self._accum_batches(train_iter)):
+            if self.gpu_verbose_level > 1:
+                logger.info("GpuRank %d: index: %d", self.gpu_rank, i)
+
+            reduce_counter += 1
```

Comment:
Is `reduce_counter` necessary? It seems that it increments exactly once per loop and is equal to i + 1.

Reply:
You are right. It is no longer useful.